// @chakra-ui/react/dist/cjs/theme-extensions/with-default-props.js
"use strict";

exports.__esModule = true;
exports.withDefaultProps = withDefaultProps;

var _utils = require("@chakra-ui/utils");

var _extendTheme = require("../extend-theme");

var _withDefaultColorScheme = require("./with-default-color-scheme");

var _withDefaultVariant = require("./with-default-variant");

var _withDefaultSize = require("./with-default-size");

function withDefaultProps(_ref) {
  var _ref$defaultProps = _ref.defaultProps,
      colorScheme = _ref$defaultProps.colorScheme,
      variant = _ref$defaultProps.variant,
      size = _ref$defaultProps.size,
      components = _ref.components;

  var identity = function identity(t) {
    return t;
  };

  var fns = [colorScheme ? (0, _withDefaultColorScheme.withDefaultColorScheme)({
    colorScheme: colorScheme,
    components: components
  }) : identity, size ? (0, _withDefaultSize.withDefaultSize)({
    size: size,
    components: components
  }) : identity, variant ? (0, _withDefaultVariant.withDefaultVariant)({
    variant: variant,
    components: components
  }) : identity];
  return function (theme) {
    return (0, _extendTheme.mergeThemeOverride)(_utils.pipe.apply(void 0, fns)(theme));
  };
}
