// @chakra-ui/utils/dist/cjs/tabbable.js
"use strict";

exports.__esModule = true;
exports.isDisabled = isDisabled;
exports.isInputElement = isInputElement;
exports.isActiveElement = isActiveElement;
exports.hasFocusWithin = hasFocusWithin;
exports.isHidden = isHidden;
exports.isContentEditable = isContentEditable;
exports.isFocusable = isFocusable;
exports.isTabbable = isTabbable;
exports.hasNegativeTabIndex = exports.hasTabIndex = exports.hasDisplayNone = void 0;

var _dom = require("./dom");

// Really great work done by Diego Haz on this one
// https://github.com/reakit/reakit/blob/master/packages/reakit-utils/src/tabbable.ts
var hasDisplayNone = function hasDisplayNone(element) {
  return window.getComputedStyle(element).display === "none";
};

exports.hasDisplayNone = hasDisplayNone;

var hasTabIndex = function hasTabIndex(element) {
  return element.hasAttribute("tabindex");
};

exports.hasTabIndex = hasTabIndex;

var hasNegativeTabIndex = function hasNegativeTabIndex(element) {
  return hasTabIndex(element) && element.tabIndex === -1;
};

exports.hasNegativeTabIndex = hasNegativeTabIndex;

function isDisabled(element) {
  return Boolean(element.getAttribute("disabled")) === true || Boolean(element.getAttribute("aria-disabled")) === true;
}

function isInputElement(element) {
  return (0, _dom.isHTMLElement)(element) && element.tagName.toLowerCase() === "input" && "select" in element;
}

function isActiveElement(element) {
  var doc = (0, _dom.isHTMLElement)(element) ? (0, _dom.getOwnerDocument)(element) : document;
  return doc.activeElement === element;
}

function hasFocusWithin(element) {
  if (!document.activeElement) return false;
  return element.contains(document.activeElement);
}

function isHidden(element) {
  if (element.parentElement && isHidden(element.parentElement)) return true;
  return element.hidden;
}

function isContentEditable(element) {
  var value = element.getAttribute("contenteditable");
  return value !== "false" && value != null;
}

function isFocusable(element) {
  if (!(0, _dom.isHTMLElement)(element) || isHidden(element) || isDisabled(element)) {
    return false;
  }

  var localName = element.localName;
  var focusableTags = ["input", "select", "textarea", "button"];
  if (focusableTags.indexOf(localName) >= 0) return true;
  var others = {
    a: function a() {
      return element.hasAttribute("href");
    },
    audio: function audio() {
      return element.hasAttribute("controls");
    },
    video: function video() {
      return element.hasAttribute("controls");
    }
  };

  if (localName in others) {
    return others[localName]();
  }

  if (isContentEditable(element)) return true;
  return hasTabIndex(element);
}

function isTabbable(element) {
  if (!element) return false;
  return (0, _dom.isHTMLElement)(element) && isFocusable(element) && !hasNegativeTabIndex(element);
}
