// @chakra-ui/react/dist/cjs/chakra-provider.js
"use strict";

exports.__esModule = true;
exports.ChakraProvider = void 0;

var _cssReset = _interopRequireDefault(require("@chakra-ui/css-reset"));

var _portal = require("@chakra-ui/portal");

var _system = require("@chakra-ui/system");

var _theme = _interopRequireDefault(require("@chakra-ui/theme"));

var _reactEnv = require("@chakra-ui/react-env");

var React = _interopRequireWildcard(require("react"));

var _hooks = require("@chakra-ui/hooks");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/**
 * The global provider that must be added to make all Chakra components
 * work correctly
 */
var ChakraProvider = function ChakraProvider(props) {
  var children = props.children,
      colorModeManager = props.colorModeManager,
      portalZIndex = props.portalZIndex,
      _props$resetCSS = props.resetCSS,
      resetCSS = _props$resetCSS === void 0 ? true : _props$resetCSS,
      _props$theme = props.theme,
      theme = _props$theme === void 0 ? _theme["default"] : _props$theme,
      environment = props.environment,
      cssVarsRoot = props.cssVarsRoot;

  var _children = /*#__PURE__*/React.createElement(_reactEnv.EnvironmentProvider, {
    environment: environment
  }, children);

  return /*#__PURE__*/React.createElement(_hooks.IdProvider, null, /*#__PURE__*/React.createElement(_system.ThemeProvider, {
    theme: theme,
    cssVarsRoot: cssVarsRoot
  }, /*#__PURE__*/React.createElement(_system.ColorModeProvider, {
    colorModeManager: colorModeManager,
    options: theme.config
  }, resetCSS && /*#__PURE__*/React.createElement(_cssReset["default"], null), /*#__PURE__*/React.createElement(_system.GlobalStyle, null), portalZIndex ? /*#__PURE__*/React.createElement(_portal.PortalManager, {
    zIndex: portalZIndex
  }, _children) : _children)));
};

exports.ChakraProvider = ChakraProvider;
